File: go_app/internal/ui/banner.go
package ui

import (
	"fmt"
	"strings"
)

// BannerVariant represents the possible variants of a banner
type BannerVariant string

const (
	// BannerVariantIntroduction represents an introduction banner
	BannerVariantIntroduction BannerVariant = "introduction"
	// BannerVariantPromotion represents a promotion banner
	BannerVariantPromotion BannerVariant = "promotion"
)

// Banner represents a banner component
type Banner struct {
	ButtonText     string
	ButtonLink     string
	Variant        BannerVariant
	SVGPath        string
	BannerOptions  map[string]interface{}
	ButtonOptions  map[string]interface{}
	CloseOptions   map[string]interface{}
	Title          string
	Illustration   string
	PrimaryAction  string
	Actions        []string
	Content        string
}

// NewBanner creates a new banner with the given parameters
func NewBanner(buttonText string, buttonLink string, variant BannerVariant, svgPath string, bannerOptions map[string]interface{}, buttonOptions map[string]interface{}, closeOptions map[string]interface{}, title string, illustration string, primaryAction string, actions []string, content string) *Banner {
	if bannerOptions == nil {
		bannerOptions = make(map[string]interface{})
	}
	if buttonOptions == nil {
		buttonOptions = make(map[string]interface{})
	}
	if closeOptions == nil {
		closeOptions = make(map[string]interface{})
	}
	if actions == nil {
		actions = []string{}
	}

	// Add default close button classes
	if _, ok := closeOptions["class"]; !ok {
		closeOptions["class"] = "js-close gl-banner-close"
	} else {
		closeOptions["class"] = fmt.Sprintf("js-close gl-banner-close %s", closeOptions["class"])
	}

	return &Banner{
		ButtonText:     buttonText,
		ButtonLink:     buttonLink,
		Variant:        variant,
		SVGPath:        svgPath,
		BannerOptions:  bannerOptions,
		ButtonOptions:  buttonOptions,
		CloseOptions:   closeOptions,
		Title:          title,
		Illustration:   illustration,
		PrimaryAction:  primaryAction,
		Actions:        actions,
		Content:        content,
	}
}

// BannerClass returns the CSS classes for the banner
func (b *Banner) BannerClass() string {
	var classes []string

	// Add variant class if needed
	if b.Introduction() {
		classes = append(classes, "gl-banner-introduction")
	}

	return strings.Join(classes, " ")
}

// CloseButtonVariant returns the variant for the close button
func (b *Banner) CloseButtonVariant() string {
	if b.Introduction() {
		return "confirm"
	}
	return "default"
}

// Introduction returns whether the banner is an introduction banner
func (b *Banner) Introduction() bool {
	return b.Variant == BannerVariantIntroduction
}

// HasIllustration returns whether the banner has an illustration
func (b *Banner) HasIllustration() bool {
	return b.Illustration != ""
}

// HasSVGPath returns whether the banner has an SVG path
func (b *Banner) HasSVGPath() bool {
	return b.SVGPath != ""
}

// HasPrimaryAction returns whether the banner has a primary action
func (b *Banner) HasPrimaryAction() bool {
	return b.PrimaryAction != ""
}

// Render generates the HTML for the banner
func (b *Banner) Render() string {
	// Build banner attributes
	bannerAttrs := make(map[string]interface{})
	for k, v := range b.BannerOptions {
		bannerAttrs[k] = v
	}

	// Add base classes
	bannerClass := "gl-banner gl-card gl-pl-6 gl-pr-8 gl-py-6"
	if bannerClass := b.BannerClass(); bannerClass != "" {
		bannerClass = fmt.Sprintf("%s %s", bannerClass, bannerClass)
	}
	bannerAttrs["class"] = bannerClass

	// Build the content
	var contentParts []string

	// Add illustration if needed
	if b.HasIllustration() {
		contentParts = append(contentParts, fmt.Sprintf(`<div class="gl-banner-illustration">%s</div>`, b.Illustration))
	} else if b.HasSVGPath() {
		contentParts = append(contentParts, fmt.Sprintf(`<div class="gl-banner-illustration"><img src="%s" alt=""></div>`, b.SVGPath))
	}

	// Add content
	contentParts = append(contentParts, `<div class="gl-banner-content">`)

	// Add title if needed
	if b.Title != "" {
		contentParts = append(contentParts, fmt.Sprintf(`<h2 class="gl-banner-title">%s</h2>`, b.Title))
	}

	// Add main content
	contentParts = append(contentParts, b.Content)

	// Add primary action or default button
	if b.HasPrimaryAction() {
		contentParts = append(contentParts, b.PrimaryAction)
	} else {
		buttonAttrs := make(map[string]interface{})
		for k, v := range b.ButtonOptions {
			buttonAttrs[k] = v
		}
		buttonAttrs["class"] = "js-close-callout"
		buttonAttrs["variant"] = "confirm"
		contentParts = append(contentParts, fmt.Sprintf(`<a href="%s" %s>%s</a>`, b.ButtonLink, formatAttributes(buttonAttrs), b.ButtonText))
	}

	// Add actions
	for _, action := range b.Actions {
		contentParts = append(contentParts, action)
	}

	contentParts = append(contentParts, `</div>`)

	// Add close button
	closeAttrs := make(map[string]interface{})
	for k, v := range b.CloseOptions {
		closeAttrs[k] = v
	}
	closeAttrs["category"] = "tertiary"
	closeAttrs["variant"] = b.CloseButtonVariant()
	closeAttrs["size"] = "small"
	closeAttrs["icon"] = "close"
	contentParts = append(contentParts, fmt.Sprintf(`<button type="button" %s><span class="js-close-icon">close</span></button>`, formatAttributes(closeAttrs)))

	// Render the banner
	return fmt.Sprintf(`<div %s><div class="gl-card-body gl-flex !gl-p-0">%s</div></div>`,
		formatAttributes(bannerAttrs),
		strings.Join(contentParts, "\n"))
}

File: rust_app/src/components/pajamas/banner.rs
use super::button::ButtonComponent;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Banner variant options
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BannerVariant {
    Introduction,
    Promotion,
}

impl BannerVariant {
    fn as_str(&self) -> &'static str {
        match self {
            BannerVariant::Introduction => "introduction",
            BannerVariant::Promotion => "promotion",
        }
    }
}

/// Renders a Pajamas compliant banner element
#[derive(Debug, Clone)]
pub struct BannerComponent {
    /// The text for the button
    button_text: String,
    /// The link for the button
    button_link: String,
    /// The variant of the banner
    variant: BannerVariant,
    /// The path to the SVG image
    svg_path: Option<String>,
    /// Options for the banner
    banner_options: HashMap<String, String>,
    /// Options for the button
    button_options: HashMap<String, String>,
    /// Options for the close button
    close_options: HashMap<String, String>,
    /// The title content
    title: Option<String>,
    /// The illustration content
    illustration: Option<String>,
    /// The primary action content
    primary_action: Option<String>,
    /// The actions content
    actions: Vec<String>,
    /// The main content
    content: Option<String>,
}

impl BannerComponent {
    /// Create a new banner component
    pub fn new(
        button_text: Option<String>,
        button_link: Option<String>,
        variant: BannerVariant,
        svg_path: Option<String>,
        banner_options: Option<HashMap<String, String>>,
        button_options: Option<HashMap<String, String>>,
        close_options: Option<HashMap<String, String>>,
    ) -> Self {
        Self {
            button_text: button_text.unwrap_or_else(|| "OK".to_string()),
            button_link: button_link.unwrap_or_else(|| "#".to_string()),
            variant,
            svg_path,
            banner_options: banner_options.unwrap_or_default(),
            button_options: button_options.unwrap_or_default(),
            close_options: Self::format_close_options(close_options),
            title: None,
            illustration: None,
            primary_action: None,
            actions: Vec::new(),
            content: None,
        }
    }

    /// Format the close button options
    fn format_close_options(options: Option<HashMap<String, String>>) -> HashMap<String, String> {
        let mut close_options = options.unwrap_or_default();
        close_options.insert("class".to_string(), "js-close gl-banner-close".to_string());
        close_options
    }

    /// Set the title content
    pub fn with_title(mut self, title: String) -> Self {
        self.title = Some(title);
        self
    }

    /// Set the illustration content
    pub fn with_illustration(mut self, illustration: String) -> Self {
        self.illustration = Some(illustration);
        self
    }

    /// Set the primary action content
    pub fn with_primary_action(mut self, primary_action: String) -> Self {
        self.primary_action = Some(primary_action);
        self
    }

    /// Add an action
    pub fn with_action(mut self, action: String) -> Self {
        self.actions.push(action);
        self
    }

    /// Set the main content
    pub fn with_content(mut self, content: String) -> Self {
        self.content = Some(content);
        self
    }

    /// Get the banner class
    fn banner_class(&self) -> String {
        if self.is_introduction() {
            "gl-banner-introduction".to_string()
        } else {
            String::new()
        }
    }

    /// Get the close button variant
    fn close_button_variant(&self) -> super::button::ButtonVariant {
        if self.is_introduction() {
            super::button::ButtonVariant::Confirm
        } else {
            super::button::ButtonVariant::Default
        }
    }

    /// Check if the banner is an introduction
    fn is_introduction(&self) -> bool {
        self.variant == BannerVariant::Introduction
    }

    /// Render the banner component
    pub fn render(&self) -> String {
        let mut html = String::new();

        // Start banner
        let banner_class = self.banner_class();
        let banner_class_str = if banner_class.is_empty() {
            String::new()
        } else {
            format!(" {}", banner_class)
        };

        let mut banner_attrs = self.banner_options.clone();
        banner_attrs.insert(
            "class".to_string(),
            format!(
                "gl-banner gl-card gl-pl-6 gl-pr-8 gl-py-6{}",
                banner_class_str
            ),
        );

        let banner_attrs_str = banner_attrs
            .iter()
            .map(|(key, value)| format!(" {}=\"{}\"", key, value))
            .collect::<Vec<String>>()
            .join("");

        html.push_str(&format!("<div{}>", banner_attrs_str));

        // Banner body
        html.push_str(r#"<div class="gl-card-body gl-flex !gl-p-0">"#);

        // Illustration
        if let Some(illustration) = &self.illustration {
            html.push_str(r#"<div class="gl-banner-illustration">"#);
            html.push_str(illustration);
            html.push_str("</div>");
        } else if let Some(svg_path) = &self.svg_path {
            html.push_str(&format!(
                r#"<div class="gl-banner-illustration"><img src="{}" alt=""></div>"#,
                svg_path
            ));
        }

        // Content
        html.push_str(r#"<div class="gl-banner-content">"#);

        // Title
        if let Some(title) = &self.title {
            html.push_str(&format!(r#"<h2 class="gl-banner-title">{}</h2>"#, title));
        }

        // Main content
        if let Some(content) = &self.content {
            html.push_str(content);
        }

        // Primary action
        if let Some(primary_action) = &self.primary_action {
            html.push_str(primary_action);
        } else {
            let mut button = ButtonComponent::new();
            button.variant = super::button::ButtonVariant::Confirm;
            button.href = Some(self.button_link.clone());
            button.content = Some(self.button_text.clone());
            button.button_options = self.button_options.clone();
            button
                .button_options
                .insert("class".to_string(), "js-close-callout".to_string());
            html.push_str(&button.render());
        }

        // Actions
        for action in &self.actions {
            html.push_str(action);
        }

        html.push_str("</div></div>");

        // Close button
        let mut close_button = ButtonComponent::new();
        close_button.category = super::button::ButtonCategory::Tertiary;
        close_button.variant = self.close_button_variant();
        close_button.size = super::button::ButtonSize::Small;
        close_button.icon = Some("close".to_string());
        close_button.button_options = self.close_options.clone();
        html.push_str(&close_button.render());

        html.push_str("</div>");
        html
    }
}
